### PR TITLE
fix: Reorder and fix reflection helpers

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
There is a mess in the code while having very similar reflection helpers for both java and kotlin. The PR puts some order there by only keeping the helpers implemented in Kotlin.

### CODE REVIEW

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/MethodUtils.java  9
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/ArrayUtils.java  7
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/ClassUtils.java  6
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/Validate.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-espresso-driver/pullRequest?prid=6237154&bid=20321334)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/Validate.java  3
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/ClassUtils.java  6
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/ArrayUtils.java  7
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/MethodUtils.java  9
         

Complexity decreasing per file
==============================
+ espresso-server/app/src/test/java/io/appium/espressoserver/test/helpers/InvocationOperationTest.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-espresso-driver/pullRequest?prid=6237154&bid=20321334)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/MethodUtils.java  9
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/Validate.java  3
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/ArrayUtils.java  7
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/ClassUtils.java  6
         

Complexity decreasing per file
==============================
+ espresso-server/app/src/test/java/io/appium/espressoserver/test/helpers/InvocationOperationTest.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-espresso-driver/pullRequest?prid=6237154&bid=20321334)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/ClassUtils.java  6
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/MethodUtils.java  9
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/Validate.java  3
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/ArrayUtils.java  7
         

Complexity decreasing per file
==============================
+ espresso-server/app/src/test/java/io/appium/espressoserver/test/helpers/InvocationOperationTest.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-espresso-driver/pullRequest?prid=6237154&bid=20321334)

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/web/WebAtomDeserializer.kt [hunk 1]
```diff
@@ -19,22 +21,27 @@ class WebAtomDeserializer : JsonDeserializer<WebAtom> {
...
                 // Validate that the locator contains "using" and "value"
-                val locator = it.asJsonObject;
+                val locator = it.asJsonObject
                 if (!locator.has("using") || !locator.has("value")) {
-                    throw JsonParseException("'locator' must have properties 'using' and 'value'");
+                    throw JsonParseException("'locator' must have properties 'using' and 'value'")
                 }
 
+                val locatorsMapping = Locator.values().fold(mutableMapOf<String, Locator>(), { acc, item ->
+                    acc[item.name.replace("_", "")] = item
```
Comment (KazuCocoa) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/web/WebAtomDeserializer.kt:41]: Is this necessary?
Possible breaking change for `CSS_SELECTOR` and `PARTIALLINKTEXT` etc

**Review by KazuCocoa — COMMENTED** [review]
I found `selectFrameByIndex` raised an error, but the previous implementation also could not handle the method. So it's okay for now.

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/web/WebAtomDeserializer.kt (hunk 1 above)
Comment (mykola-mokhnach) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/web/WebAtomDeserializer.kt:41]: I was not quite sure which strategy was applied before. Did it expect snake_case or camelCase locator type names? I don't mind implementing any of these or even both of them

**Comment by mykola-mokhnach** [comment]
> I found selectFrameByIndex raised an error

Please add the stacktrace, perhaps I could do something about it.

**Comment by KazuCocoa** [comment]
With this branch, `selectFrameByIndex` was below. (Previous implementation raise a socket hang error)

```
el = @core.wait { @driver.find_element(:class, 'android.webkit.WebView') }
        @driver.execute_script 'mobile: webAtoms', {
          webviewElement: el.ref,
          forceJavascriptEnabled: true,
          methodChain: [{
            name: 'withElement',
            atom: { name: 'selectFrameByIndex', args: 1 }  # here expect https://developer.android.com/reference/androidx/test/espresso/web/webdriver/DriverAtoms#selectframebyindex
          }]
        }
```

```
[HTTP] --> POST /wd/hub/session/1c3f97a5-66a6-459f-92b7-27fa6f34bf21/execute/sync
[HTTP] {"script":"mobile: webAtoms","args":[{"webviewElement":"d82b7ab4-e6f7-44e4-9530-91e1275be103","forceJavascriptEnabled":true,"methodChain":[{"name":"withElement","atom":{"name":"selectFrameByIndex","args":1}}]}]}
[debug] [W3C (1c3f97a5)] Calling AppiumDriver.execute() with args: ["mobile: webAtoms",[{"webviewElement":"d82b7ab4-e6f7-44e4-9530-91e1275be103","forceJavascriptEnabled":true,"methodChain":[{"name":"withElement","atom":{"name":"selectFrameByIndex","args":1}}]}],"1c3f97a5-66a6-459f-92b7-27fa6f34bf21"]
[AndroidDriver] Executing native command 'mobile: webAtoms'
[debug] [WD Proxy] Proxying [POST /appium/execute_mobile/web_atoms] to [POST http://localhost:8200/session/e8cdd421-8e64-42fd-ad02-bc32a9952b66/appium/execute_mobile/web_atoms] with body: {"webviewElement":"d82b7ab4-e6f7-44e4-9530-91e1275be103","forceJavascriptEnabled":true,"methodChain":[{"name":"withElement","atom":{"name":"selectFrameByIndex","args":1}}]}
[WD Proxy] Got response with status 500: {"id":"833327ca-38ae-47ec-81c5-9cbdfd1da00c","sessionId":"e8cdd421-8e64-42fd-ad02-bc32a9952b66","value":{"error":"unknown error","message":"java.lang.NoSuchMethodException: No such method: selectFrameByIndex() on class: androidx.test.espresso.web.webdriver.DriverAtoms","stacktrace":"java.lang.NoSuchMethodException: No such method: selectFrameByIndex() on class: androidx.test.espresso.web.webdriver.DriverAtoms\n\tat io.appium.espressoserver.lib.helpers.reflection.MethodUtils.invokeStaticMethod(MethodUtils.java:57)\n\tat io.appium.espressoserver.lib.helpers.ReflectionUtils.invokeStaticMethod(ReflectionUtils.kt:38)\n\tat io.appium.espressoserver.lib.handlers.WebAtoms.handleInternal(WebAtoms.kt:54)\n\tat io.appium.espressoserver.lib.handlers.WebAtoms.handleInternal(WebAtoms.kt:32)\n\tat io.appium.espressoserver.lib.handlers.RequestHandler$DefaultImpls.handle(RequestHandler.kt:28)\n\tat io.appium.espressoserver.lib.handlers.WebAtoms.handle(WebAtoms.kt:32)\n\tat io.appium.espressoserver.lib.handlers.WebAtoms.han...
[debug] [W3C] Matched W3C error code 'unknown error' to UnknownError
[debug] [W3C (1c3f97a5)] Encountered internal error running command: An unknown server-side error occurred while processing the command. Original error: java.lang.NoSuchMethodException: No such method: selectFrameByIndex() on class: androidx.test.espresso.web.webdriver.DriverAtoms
[debug] [W3C (1c3f97a5)] java.lang.NoSuchMethodException: No such method: selectFrameByIndex() on class: androidx.test.espresso.web.webdriver.DriverAtoms
[debug] [W3C (1c3f97a5)] 	at io.appium.espressoserver.lib.helpers.reflection.MethodUtils.invokeStaticMethod(MethodUtils.java:57)
[debug] [W3C (1c3f97a5)] 	at io.appium.espressoserver.lib.helpers.ReflectionUtils.invokeStaticMethod(ReflectionUtils.kt:38)
[debug] [W3C (1c3f97a5)] 	at io.appium.espressoserver.lib.handlers.WebAtoms.handleInternal(WebAtoms.kt:54)
[debug] [W3C (1c3f97a5)] 	at io.appium.espressoserver.lib.handlers.WebAtoms.handleInternal(WebAtoms.kt:32)
[debug] [W3C (1c3f97a5)] 	at io.appium.espressoserver.lib.handlers.RequestHandler$DefaultImpls.handle(RequestHandler.kt:28)
[debug] [W3C (1c3f97a5)] 	at io.appium.espressoserver.lib.handlers.WebAtoms.handle(WebAtoms.kt:32)
[debug] [W3C (1c3f97a5)] 	at io.appium.espressoserver.lib.handlers.WebAtoms.handle(WebAtoms.kt:32)
[debug] [W3C (1c3f97a5)] 	at io.appium.espressoserver.lib.http.Router.route(Router.kt:225)
[debug] [W3C (1c3f97a5)] 	at io.appium.espressoserver.lib.http.Server.serve(Server.kt:49)
[debug] [W3C (1c3f97a5)] 	at fi.iki.elonen.NanoHTTPD$HTTPSession.execute(NanoHTTPD.java:945)
[debug] [W3C (1c3f97a5)] 	at fi.iki.elonen.NanoHTTPD$ClientHandler.run(NanoHTTPD.java:192)
[debug] [W3C (1c3f97a5)] 	at java.lang.Thread.run(Thread.java:764)
[debug] [W3C (1c3f97a5)]
```

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/web/WebAtomDeserializer.kt (hunk 1 above)
Comment (KazuCocoa) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/web/WebAtomDeserializer.kt:41]: Ah, okay. Locator names were snake_case with upper case. https://developer.android.com/reference/androidx/test/espresso/web/webdriver/Locator

**Comment by mykola-mokhnach** [comment]
I've fixed deserialization. Should work better now

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/ClassUtils.java  6
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/Validate.java  3
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/ArrayUtils.java  7
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/MethodUtils.java  9
         

Complexity decreasing per file
==============================
+ espresso-server/app/src/test/java/io/appium/espressoserver/test/helpers/InvocationOperationTest.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-espresso-driver/pullRequest?prid=6237154&bid=20321334)

**Comment by KazuCocoa** [comment]
mm

```
[HTTP] {"script":"mobile: webAtoms","args":[{"webviewElement":"54ed0c33-3857-4c07-87db-dce9eb7a1cb0","forceJavascriptEnabled":true,"methodChain":[{"name":"withElement","atom":{"name":"selectFrameByIndex","args":1}}]}]}
[debug] [W3C (0f61ce88)] Calling AppiumDriver.execute() with args: ["mobile: webAtoms",[{"webviewElement":"54ed0c33-3857-4c07-87db-dce9eb7a1cb0","forceJavascriptEnabled":true,"methodChain":[{"name":"withElement","atom":{"name":"selectFrameByIndex","args":1}}]}],"0f61ce88-89dd-407f-8320-be0366f63cc3"]
[AndroidDriver] Executing native command 'mobile: webAtoms'
[debug] [WD Proxy] Proxying [POST /appium/execute_mobile/web_atoms] to [POST http://localhost:8200/session/43c61b06-c177-4c98-b679-5584a273043e/appium/execute_mobile/web_atoms] with body: {"webviewElement":"54ed0c33-3857-4c07-87db-dce9eb7a1cb0","forceJavascriptEnabled":true,"methodChain":[{"name":"withElement","atom":{"name":"selectFrameByIndex","args":1}}]}
[WD Proxy] Got response with status 500: {"id":"52b945d1-a652-4846-8987-d56854ba281d","sessionId":"43c61b06-c177-4c98-b679-5584a273043e","value":{"error":"unknown error","message":"java.lang.NoSuchMethodException: No such method: selectFrameByIndex() on class: androidx.test.espresso.web.webdriver.DriverAtoms","stacktrace":"java.lang.NoSuchMethodException: No such method: selectFrameByIndex() on class: androidx.test.espresso.web.webdriver.DriverAtoms\n\tat io.appium.espressoserver.lib.helpers.reflection.MethodUtils.invokeStaticMethod(MethodUtils.java:57)\n\tat io.appium.espressoserver.lib.helpers.ReflectionUtils.invokeStaticMethod(ReflectionUtils.kt:38)\n\tat io.appium.espressoserver.lib.handlers.WebAtoms.handleInternal(WebAtoms.kt:54)\n\tat io.appium.espressoserver.lib.handlers.WebAtoms.handleInternal(WebAtoms.kt:32)\n\tat io.appium.espressoserver.lib.handlers.RequestHandler$DefaultImpls.handle(RequestHandler.kt:28)\n\tat io.appium.espressoserver.lib.handlers.WebAtoms.handle(WebAtoms.kt:32)\n\tat io.appium.espressoserver.lib.handlers.WebAtoms.han...
[debug] [W3C] Matched W3C error code 'unknown error' to UnknownError
[debug] [W3C (0f61ce88)] Encountered internal error running command: An unknown server-side error occurred while processing the command. Original error: java.lang.NoSuchMethodException: No such method: selectFrameByIndex() on class: androidx.test.espresso.web.webdriver.DriverAtoms
[debug] [W3C (0f61ce88)] java.lang.NoSuchMethodException: No such method: selectFrameByIndex() on class: androidx.test.espresso.web.webdriver.DriverAtoms
[debug] [W3C (0f61ce88)] 	at io.appium.espressoserver.lib.helpers.reflection.MethodUtils.invokeStaticMethod(MethodUtils.java:57)
[debug] [W3C (0f61ce88)] 	at io.appium.espressoserver.lib.helpers.ReflectionUtils.invokeStaticMethod(ReflectionUtils.kt:38)
[debug] [W3C (0f61ce88)] 	at io.appium.espressoserver.lib.handlers.WebAtoms.handleInternal(WebAtoms.kt:54)
[debug] [W3C (0f61ce88)] 	at io.appium.espressoserver.lib.handlers.WebAtoms.handleInternal(WebAtoms.kt:32)
[debug] [W3C (0f61ce88)] 	at io.appium.espressoserver.lib.handlers.RequestHandler$DefaultImpls.handle(RequestHandler.kt:28)
[debug] [W3C (0f61ce88)] 	at io.appium.espressoserver.lib.handlers.WebAtoms.handle(WebAtoms.kt:32)
[debug] [W3C (0f61ce88)] 	at io.appium.espressoserver.lib.handlers.WebAtoms.handle(WebAtoms.kt:32)
[debug] [W3C (0f61ce88)] 	at io.appium.espressoserver.lib.http.Router.route(Router.kt:225)
[debug] [W3C (0f61ce88)] 	at io.appium.espressoserver.lib.http.Server.serve(Server.kt:49)
[debug] [W3C (0f61ce88)] 	at fi.iki.elonen.NanoHTTPD$HTTPSession.execute(NanoHTTPD.java:945)
[debug] [W3C (0f61ce88)] 	at fi.iki.elonen.NanoHTTPD$ClientHandler.run(NanoHTTPD.java:192)
[debug] [W3C (0f61ce88)] 	at java.lang.Thread.run(Thread.java:764)
[debug] [W3C (0f61ce88)]
[HTTP] <-- POST /wd/hub/session/0f61ce88-89dd-407f-8320-be0366f63cc3/execute/sync 500 71 ms - 1351
[HTTP]
```

The selector was resolved

**Comment by KazuCocoa** [comment]
Let me check the above on my local, too

**Comment by mykola-mokhnach** [comment]
np, did the necessary fix

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/ArrayUtils.java  7
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/Validate.java  3
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/MethodUtils.java  9
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/ClassUtils.java  6
         

Complexity decreasing per file
==============================
+ espresso-server/app/src/test/java/io/appium/espressoserver/test/helpers/InvocationOperationTest.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-espresso-driver/pullRequest?prid=6237154&bid=20321334)

File: espresso-server/app/src/test/java/io/appium/espressoserver/test/model/web/WebAtomsTest.kt [hunk 2]
```diff
@@ -38,7 +38,7 @@ class WebAtomsTest {
         }""".trimIndent(), WebAtomsMethod::class.java)
         assertEquals(webAtomsMethod.name, "withElement")
         assertEquals(webAtomsMethod.atom.name, "selectFrameByIndex")
-        assertEquals(webAtomsMethod.atom.args[0], 1L)
+        assertEquals(webAtomsMethod.atom.args[0], 1)
```
Comment (KazuCocoa) [inline espresso-server/app/src/test/java/io/appium/espressoserver/test/model/web/WebAtomsTest.kt:41]: 👍

**Review by KazuCocoa — APPROVED** [review]
👍 
I tested some cases, they worked

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/Validate.java  3
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/ArrayUtils.java  7
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/MethodUtils.java  9
- espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/reflection/ClassUtils.java  6
         

Complexity decreasing per file
==============================
+ espresso-server/app/src/test/java/io/appium/espressoserver/test/helpers/InvocationOperationTest.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-espresso-driver/pullRequest?prid=6237154&bid=20321334)

**Pull request merged**